File: code/python/02_ingest_disease_registry.py
"""
02_ingest_disease_registry.py
─────────────────────────────────────────────────────────────────────────────
Purpose : Ingest, validate, and clean the raw reportable disease registry.
          Key data quality issue: mixed date formats.

Input   : data/raw/reportable_disease_registry.csv
Outputs : data/clean/registry_clean.csv
          outputs/tables/dq_report_registry.csv
─────────────────────────────────────────────────────────────────────────────
"""

import os
import logging
from datetime import datetime
import numpy as np
import pandas as pd

ROOT     = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
RAW_PATH = os.path.join(ROOT, 'data', 'raw',   'reportable_disease_registry.csv')
CLN_PATH = os.path.join(ROOT, 'data', 'clean', 'registry_clean.csv')
DQ_PATH  = os.path.join(ROOT, 'outputs', 'tables', 'dq_report_registry.csv')
LOG_PATH = os.path.join(ROOT, 'outputs', 'logs',   'pipeline_run_log.txt')

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s  %(levelname)-8s  %(message)s',
    handlers=[logging.FileHandler(LOG_PATH, mode='a'), logging.StreamHandler()]
)
log = logging.getLogger(__name__)
log.info("=" * 60)
log.info("START: 02_ingest_disease_registry.py")
log.info("=" * 60)

df = pd.read_csv(RAW_PATH, dtype=str)
log.info(f"Loaded {len(df):,} raw rows")

dq = []

# ── 1. Standardize dates (mixed formats) ────────────────────────────────────
def parse_date_mixed(s):
    """Try ISO 8601 first, then MM/DD/YYYY."""
    if pd.isna(s):
        return pd.NaT
    for fmt in ('%Y-%m-%d', '%m/%d/%Y', '%d-%m-%Y'):
        try:
            return datetime.strptime(str(s).strip(), fmt)
        except ValueError:
            continue
    return pd.NaT

df['report_date_parsed'] = df['report_date'].apply(parse_date_mixed)
n_bad_dates = df['report_date_parsed'].isna().sum()
dq.append({'check': 'Unparseable report_date', 'n_affected': n_bad_dates,
           'action': 'Set to NULL'})
log.info(f"Date parsing: {len(df) - n_bad_dates}/{len(df)} parsed successfully")

# Count how many were in non-ISO format
n_non_iso = df['report_date'].str.contains('/', na=False).sum()
dq.append({'check': 'Non-ISO date formats (MM/DD/YYYY)', 'n_affected': int(n_non_iso),
           'action': 'Parsed and converted to ISO 8601'})

df['report_year'] = df['report_date_parsed'].apply(
    lambda d: d.year if pd.notna(d) else np.nan
)

# ── 2. Patient age ────────────────────────────────────────────────────────────
df['patient_age_num'] = pd.to_numeric(df['patient_age'], errors='coerce')
n_miss_age = df['patient_age_num'].isna().sum()
dq.append({'check': 'Missing patient_age', 'n_affected': int(n_miss_age),
           'action': 'Retained as NULL; flagged'})
df['age_missing'] = df['patient_age_num'].isna().astype(int)

# ── 3. Sex ────────────────────────────────────────────────────────────────────
valid_sex = {'Male', 'Female', 'Unknown'}
n_invalid_sex = (~df['patient_sex'].isin(valid_sex)).sum()
if n_invalid_sex:
    dq.append({'check': 'Invalid sex values', 'n_affected': int(n_invalid_sex),
               'action': 'Set to Unknown'})
    df['sex_clean'] = df['patient_sex'].where(df['patient_sex'].isin(valid_sex), 'Unknown')
else:
    df['sex_clean'] = df['patient_sex']
    dq.append({'check': 'Sex values', 'n_affected': 0, 'action': 'All valid — no action'})

# ── 4. Outbreak indicator ─────────────────────────────────────────────────────
df['is_outbreak_case'] = (~df['outbreak_id'].isna() & (df['outbreak_id'] != '')).astype(int)

# ── 5. Hospitalization and death binary ──────────────────────────────────────
for col, new_col in [('hospitalized', 'hospitalized_bin'), ('died', 'died_bin')]:
    df[new_col] = df[col].map({'Yes': 1, 'No': 0}).fillna(np.nan)

# ── 6. County FIPS zero-padding ───────────────────────────────────────────────
df['county_fips_clean'] = df['county_fips'].str.strip().str.zfill(5)

# ── 7. Condition name standardization ────────────────────────────────────────
df['condition_name_clean'] = df['condition_name'].str.strip().str.title()

# ── Save ──────────────────────────────────────────────────────────────────────
KEEP = [
    'case_id', 'report_date_parsed', 'report_year', 'condition_name_clean',
    'icd10_code', 'patient_age_num', 'sex_clean', 'county_fips_clean',
    'outbreak_id', 'is_outbreak_case', 'hospitalized', 'hospitalized_bin',
    'died', 'died_bin', 'source_lab', 'age_missing',
]
df_clean = df[KEEP].copy()
df_clean.columns = [
    'case_id', 'report_date', 'report_year', 'condition_name',
    'icd10_code', 'patient_age', 'sex', 'county_fips',
    'outbreak_id', 'is_outbreak_case', 'hospitalized', 'hospitalized_bin',
    'died', 'died_bin', 'source_lab', 'age_missing',
]
df_clean.to_csv(CLN_PATH, index=False)
log.info(f"Clean registry saved: {len(df_clean):,} rows")

pd.DataFrame(dq).assign(dataset='reportable_disease_registry',
                         run_timestamp=datetime.now().strftime('%Y-%m-%d %H:%M:%S')
                        ).to_csv(DQ_PATH, index=False)
log.info("END: 02_ingest_disease_registry.py")
print(f"\n✓ Registry cleaning complete. {len(df_clean):,} records saved.")

File: code/r/02_ingest_disease_registry.R
# ============================================================================
# 02_ingest_disease_registry.R
# Purpose: Clean the reportable disease registry; handle mixed date formats.
# Input:  data/raw/reportable_disease_registry.csv
# Output: data/clean/registry_clean_r.csv
# ============================================================================

library(tidyverse)
library(lubridate)

PROJECT_ROOT <- here::here()

raw_path <- file.path(PROJECT_ROOT, "data", "raw",   "reportable_disease_registry.csv")
cln_path <- file.path(PROJECT_ROOT, "data", "clean", "registry_clean_r.csv")

df_raw <- read_csv(raw_path, col_types = cols(.default = "c"), show_col_types = FALSE)
message(sprintf("Loaded %s registry rows", format(nrow(df_raw), big.mark = ",")))

# ── Parse mixed date formats ───────────────────────────────────────────────────
parse_mixed_date <- function(x) {
  # Try ISO 8601 first, then MM/DD/YYYY
  d <- parse_date_time(x, orders = c("Ymd", "mdy"), quiet = TRUE)
  as.Date(d)
}

df <- df_raw %>%
  mutate(
    report_date_parsed = parse_mixed_date(report_date),
    report_year        = year(report_date_parsed),
    patient_age_num    = suppressWarnings(as.numeric(patient_age)),
    age_missing        = as.integer(is.na(patient_age_num)),
    is_outbreak_case   = as.integer(!is.na(outbreak_id) & outbreak_id != ""),
    hospitalized_bin   = case_when(
      hospitalized == "Yes" ~ 1L,
      hospitalized == "No"  ~ 0L,
      TRUE ~ NA_integer_
    ),
    died_bin = case_when(
      died == "Yes" ~ 1L,
      died == "No"  ~ 0L,
      TRUE ~ NA_integer_
    ),
    county_fips_clean    = str_pad(str_trim(county_fips), 5, "left", "0"),
    condition_name_clean = str_to_title(str_trim(condition_name))
  )

n_bad_dates <- sum(is.na(df$report_date_parsed))
message(sprintf("  Date parsing: %d/%d successfully parsed", nrow(df) - n_bad_dates, nrow(df)))
message(sprintf("  Mixed format dates (MM/DD/YYYY): %d",
                sum(str_detect(df_raw$report_date, "/"), na.rm = TRUE)))

df_clean <- df %>%
  select(
    case_id,
    report_date     = report_date_parsed,
    report_year,
    condition_name  = condition_name_clean,
    icd10_code,
    patient_age     = patient_age_num,
    sex             = patient_sex,
    county_fips     = county_fips_clean,
    outbreak_id,
    is_outbreak_case,
    hospitalized,
    hospitalized_bin,
    died,
    died_bin,
    source_lab,
    age_missing
  )

write_csv(df_clean, cln_path)
message(sprintf("✓ Registry clean: %s rows → %s", format(nrow(df_clean), big.mark = ","), cln_path))
